lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -40,6 +49,52 @@ activityForEpoch epoch s =
   where
     slotInCurrentEpoch = ((epoch ==) . epochNumber)
 
+data ErrMetricsInconsistency
+    = ErrMetricsInconsistencyBlockProducerNotInStakeDistr PoolId Int
+    deriving (Show, Eq)
+
+-- | Combines two different sources of data into one:
+--
+-- 1. A stake-distribution map
+-- 2. A pool-production map
+--
+-- If a pool has produced a block without existing in the stake-distribution,
+-- i.e it exists in (2) but not (1), this function will return
+-- @Left ErrMetricsInconsistencyBlockProducerNotInStakeDistr@.
+--
+-- If a pool is in (1) but not (2), it simply means it has produced 0 blocks so
+-- far.
+combineMetrics
+    :: Map PoolId (Quantity "lovelace" Word64)
+    -> Map PoolId Int
+    -> Either
+        ErrMetricsInconsistency
+        (Map PoolId (Quantity "lovelace" Word64, Int))
+combineMetrics =
+    mergeA
+        stakeButNoActivity
+        activityButNoStake
+        (zipWithMatched (\_k stake productions -> (stake, productions)))
```

Comment:
I like how the first two cases are made very clear with the intermediary names. I'd have carried on for the 3rd case too:

```hs
mergeA stakeButNoActivity activityButNoStake stakeAndActivity
```